File: AlgoExpertIo/AlgoExpertIo/riverSizes.py
#O(w*h) time
#O(w*h) same space.
def riverSizes(matrix):
    sizes = []
    visited = [[False for value in row] for row in matrix]
    print(visited)
    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            #if node has True for visited... skip it.
            if(visited[i][j]):
                continue
            traverseNode(i, j, matrix, visited, sizes)
    return sizes 

def traverseNode(i, j, matrix, visited, sizes):
    currentRiverSize = 0
    nodesToExplore =[[i, j]] #explore the node at i j (essentially x and y coordinates for the location of the value)
    while len(nodesToExplore):
        currentNode = nodesToExplore.pop()
        i = currentNode[0]
        j = currentNode[1]

        if (visited[i][j]): #if i,j has been visited...
            continue
        visited[i][j] = True # else, set it to visited in the visited matrix.

        if(matrix[i][j]==0): # if the value is 0, then it's land, so skip it.
            continue
        currentRiverSize +=1 #else, it's a 1 for river, so expand the current river size.
        unvisitedNeighbors = getUnvisitedNeighbors(i, j, matrix, visited) #then get unvisited neighbors.
        for neighbor in unvisitedNeighbors:
            nodesToExplore.append(neighbor)

        if currentRiverSize > 0:
            sizes.append(currentRiverSize)

def getUnvisitedNeighbors(i, j, matrix, visited):
    unvisitedNeighbors = []
    if i > 0 and not visited[i -1][j]:
        unvisitedNeighbors.append([i-1, j])
    if i < len(matrix) -1 and not visited[i+1][j]:
        unvisitedNeighbors.append([i+1, j])
    if j > 0 and not visited [i][j-1]:
        unvisitedNeighbors.append([i, j -1])
    if j < len(matrix[0]) - 1 and not visited[i][j +1]:
        unvisitedNeighbors.append([i, j+1])
    return unvisitedNeighbors

# riverSizes([[1,0,0,0],[1,0,1,0], [0,1,1,1]])
print(riverSizes([[1,0,0,0],[1,0,1,0], [0,1,1,1]]))
# [1,0,0,0] #2, 4
# [1,0,1,0] 
# [0,1,1,1]

File: AlgoExpertIo/AlgoExpertIo/src/main/java/com/algoexpert/algoexpert/riversizes/RiverSizes.java
package com.algoexpert.algoexpert.riversizes;

import java.util.ArrayList;
import java.util.List;

public class RiverSizes {

    public static List<Integer> riverSizes(int[][] matrix){
        // initialize a list for storing the river sizes
        // initialize a matrix of boolean values to record visited values

        List<Integer> sizesList = new ArrayList<>();
        boolean[][] visited = new boolean[matrix.length][matrix[0].length];
        for (int i = 0; i < visited.length; i++) {
            for (int j = 0; (j) < visited[0].length ; (j)++) {
                // if node has been visited, then don't bothwr.
                if(visited[i][j]){
                    continue;
                }
                // otherwise, let's traverse it's neighbors
                traverseNode(i, j, matrix, visited, sizesList);
            }

        }

        return sizesList;
    }

    public static void traverseNode(int i, int j, int[][] matrix, boolean[][] visited,
                                    List<Integer> sizes){
        int currentRiverSize = 0;
        // each item in the yetToExplore list is an x,y or i,j coordinate.
        // it corresponds to the coordinates of a value we have not yet visited.
        List<Integer[]> yetToExplore = new ArrayList<Integer[]>();
        yetToExplore.add(new Integer[]{i,j});
        while (!yetToExplore.isEmpty()) {
            /** Extract the i,j coordinates as 'currentNode' */
            Integer[] currentNode = yetToExplore.get(yetToExplore.size()-1);
            /** Remove the item from yet to explore*/
            yetToExplore.remove(yetToExplore.size()-1);
            /** Extract the 'i' into i, 'j' into j*/
            i = currentNode[0];
            j = currentNode[1];
            /** Check if it's been visited. SKip it if it has.*/
            if (visited[i][j]){
                continue;
            }
            /** Else, visit and record our visit to this i,j coordinate */
            visited[i][j] = true;

            /** If the corresponding position of the matrix contains 0, then don't traverse it's neighbors. */
            if (matrix[i][j] == 0){
                continue;
            }
            /** Otherwise, increment the size of curent river. */
            currentRiverSize++;
            /** Determine how many of the current node's neighbors have been visited*/
            List<Integer[]> unvisitedNeighbors = getUnvisitedNeighbors(i, j, matrix, visited);
            /** Traverse the unvisited neighbors and add to yetToExplore*/
            for (Integer[] neighbor : unvisitedNeighbors) {
                yetToExplore.add(neighbor);
            }
            /** Add the river size to the sizes list.*/
            if (currentRiverSize > 0) {
                sizes.add(currentRiverSize);
            }
        }
    }

    public static List<Integer[]> getUnvisitedNeighbors(int i, int j, int[][] matrix, boolean[][] visited){
        List<Integer[]> unvisitedNeighbors = new ArrayList<Integer[]>();

        /** i corresponds to the row, j to the column. So below, we look above, below, left and right of
         * the current node at i, j and determine if it's been visited.
         * If it has not been, then we add it to the
         * */
        if (i > 0 & !visited[i-1][j]){
            unvisitedNeighbors.add(new Integer[]{i - 1, j});
        }
        if (i < matrix.length - 1 && !visited[i+1][j]){
            unvisitedNeighbors.add(new Integer[]{i + 1, j});
        }
        if (j > 0 && !visited[i][j - 1]) {
            unvisitedNeighbors.add(new Integer[]{i, j - 1});
        }
        if (j < matrix[0].length && !visited[i][j + 1]) {
            unvisitedNeighbors.add(new Integer[]{i, j + 1});
        }

        return unvisitedNeighbors;
    }
}

